lib/pacer_test.go

```diff
@@ -0,0 +1,148 @@
+		{2, time.Second, (49 * time.Second) / 10, 9, 100 * time.Millisecond, false},
+	} {
+		r := &Rate{Freq: tt.freq, Per: tt.per}
+
+		wait, stop := r.Pace(tt.elapsed, tt.count)
+		if have, want := wait, tt.wait; have != want {
+			t.Errorf("test case %d: %+v: wait=%v, want %v", i, r, have, want)
+		}
+		if have, want := stop, tt.stop; have != want {
+			t.Errorf("test case %d: %+v: stop=%v, want %v", i, r, have, want)
+		}
+	}
+}
+
+var quarterPeriods = map[string]float64{
+	"MeanUp":   MeanUp,
+	"Peak":     Peak,
+	"MeanDown": MeanDown,
+	"Trough":   Trough,
+}
+
+type sineTest struct {
+	p, m, a int
+}
+
+func (st sineTest) Pacer(startAt float64) SinePacer {
+	return SinePacer{
+		Period:  time.Duration(st.p) * time.Second,
+		Mean:    Rate{st.m, time.Second},
+		Amp:     Rate{st.a, time.Second},
+		StartAt: startAt,
+	}
+}
+
+func (st sineTest) AmpHits() float64 {
+	return float64(st.a*st.p) / (2 * math.Pi)
+}
+
+func (st sineTest) Hits(frac, startAt float64) uint64 {
+	return uint64(math.Round(
+		float64(st.m*st.p)*frac +
+			st.AmpHits()*(math.Cos(startAt)-math.Cos(startAt+frac*2*math.Pi))))
+}
+
+func (st sineTest) Nanos(frac, startAt float64) time.Duration {
+	return time.Duration(1 / (float64(st.m) + float64(st.a)*math.Sin(startAt+frac*2*math.Pi)))
+}
+
+func TestSinePacerHits(t *testing.T) {
+	tests := []sineTest{
+		{20 * 60, 100, 90},
+		{60, 1000, 10},
+		{1, 10, 7},
+		{1, 1, 0},
+		// These test cases failed with off-by-one errors before applying
+		// math.Round in Hits, due to floating-point maths differences.
+		{1e6, 10, 7},
+		{60, 1000, 999},
+	}
+
+	for i, test := range tests {
+		for name, sa := range quarterPeriods {
+			sr := test.Pacer(sa)
+			if got, want := sr.Hits(sr.Period/4), test.Hits(0.25, sa); got != want {
+				t.Errorf("%d(%s): hits after 1/4 period = %d, want %d", i, name, got, want)
+			}
+			if got, want := sr.Hits(sr.Period/2), test.Hits(0.5, sa); got != want {
+				t.Errorf("%d(%s): hits after 1/2 period = %d, want %d", i, name, got, want)
+			}
+			if got, want := sr.Hits(3*sr.Period/4), test.Hits(0.75, sa); got != want {
+				t.Errorf("%d(%s): hits after 3/4 period = %d, want %d", i, name, got, want)
+			}
+			if got, want := sr.Hits(sr.Period), test.Hits(1, sa); got != want {
+				t.Errorf("%d(%s): hits after full period = %d, want %d", i, name, got, want)
+			}
+		}
+	}
+}
+
+func TestSinePacerFlat(t *testing.T) {
+	st := sineTest{1, 1, 0}
+	tests := []struct {
+		et   time.Duration
+		c    uint64
+		wait time.Duration
+		stop bool
+	}{
+		{0, 0, time.Second, false},
+		{0, 1, 2 * time.Second, false},
+		{time.Second / 100, 0, 99 * time.Second / 100, false},
+		{time.Second / 2, 0, time.Second / 2, false},
+		{64 * time.Second / 100, 0, 36 * time.Second / 100, false},
+		// Has an off-by-one I can't round away nicely because it's
+		// due to expectedHits being 0.9900000000000001. Ugh floats.
+		// {99 * time.Second / 100, 0, time.Second / 100},
```

Comment:
What about comparing floats "to a certain degree" (say, using https://github.com/google/go-cmp/tree/master/cmp) ?

I think, it is the only reasonable approach ...

Reply:
Thank you, that's absolutely what I should have been doing :-)